File: supCPM/cdist.m
function Dst = cdist(D,no_dims,compel_force)
% The function computes the Capacity adjusted distance based on the pairwise
% distance in the original metric
% D: Pairwise distance matrix
% no_dims: mapped dimension
% (optional) w (=2 or 3): determines the dimension correction power 
     n = size(D,1);   
     nscales = 7; % number of scales for dimension estimation
     [dim1,r] = compute_dim(sqrt(D),nscales,0.01);
     if compel_force == 1
         dim1(2) = 1.5*dim1(2);
     end
     max_dim = 10*(no_dims-1);   % set an upper bound on the estimated dimension to avoid divergence when the data dimension is too big
     ind = (dim1<max_dim);
     dim = ones(size(dim1))*max_dim;
     dim(ind) = dim1(ind);
     dim(dim1<no_dims) = no_dims; %the lower bound on the estimated dimension is set to the mapped-to dimension
    
    % for i=1:nscales-1
    %     dim(i+1)=max(dim(i),dim(i+1));
    % end
     r1 = prctile(sort(sqrt(D(:))),1); % a small postive number preventing dividing by 0 when defining the probability
     [~, I] = sort(D(:),'ascend');
     Dst = zeros(n,n);
     for  i = 1:n
         for j = 1:n
          curr_dim = find_dim(r,dim,sqrt(D(i,j))); % find the estimated dimension at the scale r
          Dst(i,j) = (r1^2+(D(i,j)))^((curr_dim)/(no_dims));   % defined the new distance
         end
     end
 
      Dst(I) =  make_mono(sqrt(Dst(I)),sqrt(D(I)+r1^2));  % impose monotonicity
     Dst = sqrt(Dst);
end
function d = find_dim(r,dim,a)
% find the estimated dimension at the scale r
i = 1;
while i<= length(r)
    if a <= r(i)
        d = dim(i);
        break;
    else
        i = i+1;
    end
end
end
function x=make_mono(x,y)
d = max(x(1)-y(1),0);
for i = 2:length(x)
    x(i) = max([x(i),x(i-1),y(i)+d]);
    d = max(d,max(x(i)-y(i),0));
end
x = x.^2;
end

File: supCPM/R/cdist.R
cdist <- function(D,no_dims,compel_force,epsilon=1){
  n <- nrow(D)
  nscale <- 7
  dim <- compute.dim(sqrt(D),nscale,0.01)
  r <- dim$r
  dim1 <- dim$dim
  if (compel_force == 1){
    dim1[2] <- 1.5 * dim1[2]
  }
  max_dim <- 10*(no_dims-1)
  ind <- (dim1 < max_dim)
  dim <- rep(1,length(dim1)) * max_dim
  dim[ind] <- dim1[ind]
  dim[dim1<no_dims] <- no_dims

  r1 = quantile(sqrt(D),0.01,type=5)*epsilon
  I = order(D)
#  Dst <- matrix(rep(0,n^2),n)
#  for(i in 1:n){
#    for(j in 1:n){
#        curr_dim = find.dim(r,dim,sqrt(D[i,j])) # find the estimated dimension at the scale r
#    }
#  }
  curr_dim = matrix(0,n,n)
  r0 = c(-1,r)
  for (k in 2:(length(r)+1)){
    curr_dim[sqrt(D)<=r0[k]&sqrt(D)>r0[k-1]] = dim[k-1]
  }
  Dst = (r1^2+D)^(curr_dim/no_dims)
  Dst[I] = make.mono(sqrt(Dst[I]),sqrt(D[I]+r1^2))
  Dst = sqrt(Dst)
  return(Dst)
}


find.dim <- function(r,dim,a){
  i = 1
  while (i <= length(r)){
    if (a <= r[i]){
      d <- dim[i]
      break
    }else{
      i <- i+1
    }
  }
  return(d)
}

make.mono <- function(x,y){
  d <- max(x[1]-y[1],0)
  for (i in 2:length(x)){
    x[i] <- max(x[i],x[i-1],y[i]+d)
    d <- max(d,max(x[i]-y[i],0))
  }
  x = x^2
  return(x)
}


